### PR TITLE
idp/usso: support redirect login flow

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/73/

**Comment by mhilton** [comment]
Protocol Description
====
Here's a quick summary of the implemented redirect protocol. This will be documented properly when it's finalized but it might help in understanding the PR.

### Discovery
Through some (currently unimplemented) means the client gets two URLS `$login_url` and `$token_url`.

### Login
The client redirects the browser to `$login_url?return_to=$return_url[&state=$return_state]` Where the return URL is the location candid should return to when the login attempt is finished and state contains optional state information for the client. If set state is returned unchanged in the result.

### Successful login
On a successful login candid responds by redirecting to `$return_url?code=$access_code[&state=$return_state]`. The code is a code that the client can the use to request the discharge token from candid.  The client gets the actual discharge token by POSTing the code to $token_url in a body like:

```
{
    "code": "$access_code" 
}
```

### Failed login
On failure candid responds by redirecting to `$return_url?error=$error_message[&error_code=$error_code][&state=$return_state]`. error_message and (possibly) error code describing the failure.

File: idp/idputil/secret/codec.go [hunk 1]
```diff
@@ -97,3 +112,38 @@ func (c *Codec) decrypt(out, in []byte) ([]byte, error) {
...
+
+// SetCookie encodes the given value as a session cookie with the given
+// name. The returned value should be used to verify the cookie when
+// later decoding it.
+func (c *Codec) SetCookie(w http.ResponseWriter, name string, v interface{}) (string, error) {
+	out, err := c.encode(v)
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	hash := sha256.Sum224(out)
```
Comment (rogpeppe) [inline idp/idputil/secret/codec.go]: I think I'd probably use SHA256 - it's only 4 bytes more when encoded; not sure that it's worth using the smaller, slightly less standard, hash.

File: idp/idputil/secret/codec.go [hunk 2]
```diff
@@ -97,3 +112,38 @@ func (c *Codec) decrypt(out, in []byte) ([]byte, error) {
 	}
 	return out, nil
 }
+
+// SetCookie encodes the given value as a session cookie with the given
+// name. The returned value should be used to verify the cookie when
```
Comment (rogpeppe) [inline idp/idputil/secret/codec.go]: // The returned value is used the verify the cookie later - it should be passed to Cookie when the cookie is retrieved.

?

File: internal/discharger/discharge.go [hunk 3]
```diff
@@ -236,3 +237,27 @@ func newDischargeID() (string, error) {
...
+
+type dischargeTokenRequestBody struct {
+	Code string `json:"code"`
+}
+
+type dischargeTokenResponse struct {
+	DischargeToken *httpbakery.DischargeToken `json:"token,omitempty"`
+}
+
+func (h *handler) DischargeToken(p httprequest.Params, req *dischargeTokenRequest) (*dischargeTokenResponse, error) {
```
Comment (rogpeppe) [inline internal/discharger/discharge.go]: doc comment please.

File: internal/discharger/idp.go [hunk 4]
```diff
@@ -144,3 +148,56 @@ func (c *visitCompleter) Failure(ctx context.Context, w http.ResponseWriter, req
 	}
 	identity.WriteError(ctx, w, err)
 }
+
+func (c *visitCompleter) RedirectSuccess(ctx context.Context, w http.ResponseWriter, req *http.Request, returnTo, state string, id *store.Identity) {
```
Comment (rogpeppe) [inline internal/discharger/idp.go]: doc comment please.

File: internal/discharger/idp.go [hunk 5]
```diff
@@ -144,3 +148,56 @@ func (c *visitCompleter) Failure(ctx context.Context, w http.ResponseWriter, req
...
+	if state != "" {
+		v.Set("state", state)
+	}
+	if err := redirect(w, req, returnTo, v); err != nil {
+		identity.WriteError(ctx, w, err)
+	}
+	return
+}
+
+func (c *visitCompleter) RedirectFailure(ctx context.Context, w http.ResponseWriter, req *http.Request, returnTo, state string, err error) {
```
Comment (rogpeppe) [inline internal/discharger/idp.go]: doc comment please.

File: internal/discharger/idp.go [hunk 6]
```diff
@@ -144,3 +148,56 @@ func (c *visitCompleter) Failure(ctx context.Context, w http.ResponseWriter, req
...
+	if ec, ok := errgo.Cause(err).(params.ErrorCode); ok {
+		v.Set("error_code", string(ec))
+	}
+	if rerr := redirect(w, req, returnTo, v); rerr == nil {
+		return
+	}
+	identity.WriteError(ctx, w, err)
+}
+
+func redirect(w http.ResponseWriter, req *http.Request, returnTo string, values url.Values) error {
```
Comment (rogpeppe) [inline internal/discharger/idp.go]: I'd find a doc comment useful here too.

File: internal/discharger/internal/store.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+// only be available in the store until the given expire time.
+func (s *DischargeTokenStore) Put(ctx context.Context, dt *httpbakery.DischargeToken, expire time.Time) (string, error) {
+	entry := dischargeTokenEntry{
+		DischargeToken: dt,
+		Expire:         expire,
+	}
+	b, err := json.Marshal(entry)
+	if err != nil {
+		// This should be impossible
+		panic(errgo.Mask(err))
```
Comment (rogpeppe) [inline internal/discharger/internal/store.go]: not sure it's worth masking tbh.

File: internal/discharger/internal/store_test.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+type testSetStore struct {
+	store.KeyValueStore
+	f func(context.Context, string, []byte, time.Time) error
+}
+
+func (s testSetStore) Set(ctx context.Context, key string, value []byte, expire time.Time) error {
+	return s.f(ctx, key, value, expire)
+}
+
+func withSet(store store.KeyValueStore, f func(context.Context, string, []byte, time.Time) error) store.KeyValueStore {
```
Comment (rogpeppe) [inline internal/discharger/internal/store_test.go]: a doc comment might be nice here. or maybe just rename `f` to `set` ?

File: internal/discharger/internal/store_test.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+type testGetStore struct {
+	store.KeyValueStore
+	f func(context.Context, string) ([]byte, error)
+}
+
+func (s testGetStore) Get(ctx context.Context, key string) ([]byte, error) {
+	return s.f(ctx, key)
+}
+
+func withGet(store store.KeyValueStore, f func(context.Context, string) ([]byte, error)) store.KeyValueStore {
```
Comment (rogpeppe) [inline internal/discharger/internal/store_test.go]: a doc comment might be nice here. or maybe just rename f to get ?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/74/

File: idp/idputil/secret/codec.go [hunk 10]
```diff
@@ -97,3 +112,38 @@ func (c *Codec) decrypt(out, in []byte) ([]byte, error) {
...
+func (c *Codec) Cookie(req *http.Request, name, verification string, v interface{}) error {
+	cookie, err := req.Cookie(name)
+	if err != nil {
+		return errgo.WithCausef(err, ErrInvalidCookie, "invalid cookie")
+	}
+	buf, err := base64.URLEncoding.DecodeString(cookie.Value)
+	if err != nil {
+		return errgo.WithCausef(err, ErrInvalidCookie, "invalid cookie")
+	}
+	hash := sha256.Sum224(buf)
```
Comment (fabricematrat) [inline idp/idputil/secret/codec.go]: why 224 ?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/candid/75/

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.theblues.io:8080/job/candid-merge

**Pull request merged**